apps/charterafrica/src/payload/fields/ecosystem/airtableColumnSelect.js

```diff
@@ -0,0 +1,78 @@
+import { deepmerge } from "@mui/utils";
+import {
+  Select,
+  getSiblingData,
+  useAllFormFields,
+} from "payload/components/forms";
+import { select } from "payload/dist/fields/validations";
+import { createElement, useMemo } from "react";
+import useSWR from "swr";
+
+import fetchJson from "../../../utils/fetchJson";
+
+function getUrl(baseId) {
+  if (baseId) {
+    return `${process.env.PAYLOAD_PUBLIC_APP_URL}/api/v1/resources/ecosystem/schema?source=airtable&url=/meta/bases/${baseId}/tables`;
+  }
+  return null;
+}
+
+function fieldsToOptions(fields) {
+  return (
+    fields?.map((item) => ({
+      value: item.name,
+      label: item.name,
+      id: item.id,
+    })) || []
+  );
+}
+
+export function validateColumnSelect(tableField) {
+  return async function validate(value, { hasMany, required, t, data }) {
+    const url = getUrl(data?.baseId);
+    if (url) {
+      const tableId = data.schema[tableField];
+      const { tables } = await fetchJson.get(url);
+      const table = tables?.find(({ id }) => id === tableId);
+      const options = fieldsToOptions(table?.fields);
+      return select(value, { hasMany, options, required, t });
+    }
+    return [];
+  };
+}
+
+const ColumnSelect = (tableField) => {
+  return function CSelect(props) {
+    const [fields] = useAllFormFields();
+    const document = getSiblingData(fields, "schema");
+    const { baseId, schema } = document;
+    const tableId = schema[tableField];
+    const { data = {} } = useSWR(getUrl(baseId), fetchJson.get);
+    const table = data?.tables?.find(({ id }) => id === tableId);
+    const options = useMemo(
+      () => fieldsToOptions(table?.fields),
+      [table?.fields]
+    );
+
+    return createElement(Select, { ...props, options });
+  };
+};
+
+function airtableColumnSelect({ tableField, overrides = {} }) {
+  const defaultSelect = {
+    type: "text",
+    validate: validateColumnSelect(tableField),
```

Comment:
@kilemensi we have a very interesting limitation on this.

it is being called on save for every field therefore exceeding airtable api limit. We may need to find a way of caching airtable response.

Reply:
🤯 ... makes sense though @koechkevin .

My initial thinking is we should have an hidden field, `baseData` or `baseTables` where by once a user has selected the base, we make a single API request and store the tables in this field. The rest of the fields in `schema` would then refer to this data i.e. siblingData.baseTables instead of making API requests.

i.e. only one extra request for every change in base selection.

Reply:
> 🤯 ... makes sense though @koechkevin .
> 
> My initial thinking is we should have an hidden field, `baseData` or `baseTables` where by once a user has selected the base, we make a single API request and store the tables in this field. The rest of the fields in `schema` would then refer to this data i.e. `siblingData.baseTables` instead of making API requests.
> 
> i.e. only one extra request for every change in base selection.

Looks like the payloadcms api doesn't offer a way to mutate sibling data inside field hooks.

Tried in memory cache that expires after 5 minutes and it looks to work.

i.e 
```js
const cache = {};

function getFromCache(key) {
  const rawData = cache[key];
  if (rawData) {
    const { lastUpdated, value } = rawData;
    const fiveMinutesAgo = new Date();
    fiveMinutesAgo.setMinutes(fiveMinutesAgo.getMinutes() - 5);
    if (fiveMinutesAgo < lastUpdated) {
      return value;
    }
  }
  return null;
}

export async function schema(req) {
  const { url } = req.query;
  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${process.env.AIRTABLE_API_TOKEN}`,
  };
  if (getFromCache(url)) {
    return getFromCache(url);
  }
  const value = fetchJson.get(`https://api.airtable.com/v0${url}`, {
    headers,
  });
  cache[url] = { value, lastUpdated: new Date() };
  return value;
}
```

Reply:
Cool, lets implement this approach for now @koechkevin but I think the approach of may be storing cache data in the public folder is may be best moving forward (we do this in the codeforafrica app).

Create an issue to track this so that we can revisit once we've deployed current implementation